### ISSUE TITLE
Feature: #[trace("a-span", black-box=true)]

### ISSUE THREAD

**Issue by taqtiqa-mark** [tikv/minitrace-rust#122]
This feature request arose in the course of iterating on an integration test for the proc-macro-attribute `#[trace(...)]`.

```rust
use minitrace::trace;
use regex::Regex;
use test_utilities::*;

#[trace("test-span")]
fn f(a: u32) -> u32 {
    a
}

fn main() {
    let (root, collector) = minitrace::Span::root("root");
    {
        root.set_local_parent();
        f(1);
    }
    drop(root);
    let records: Vec<minitrace::collector::SpanRecord> =
        futures::executor::block_on(collector.collect());
    let expected = r#"[ ... ]"#;
    let pre = format!("{records:#?}");
    let re1 = Regex::new(r"begin_unix_time_ns: \d+,").unwrap();
    let re2 = Regex::new(r"duration_ns: \d+,").unwrap();
    let int: std::string::String = re1.replace_all(&pre, r"begin_unix_time_ns: \d+,").into();
    let actual: std::string::String = re2.replace_all(&int, r"duration_ns: \d+,").into();
    assert_eq_text!(expected, &actual);
}

```

The black-box feature [requires nightly](https://doc.rust-lang.org/std/hint/fn.black_box.html).

Requirements:
- The macro should report the nightly requirement in any error messages related to `black-box=true`.
- The default is `black-box=false`

**Comment by andylokandy** [comment]
Could you give more context about why `black_box` is needed?

**Comment by taqtiqa-mark** [comment]
Originally I thought issue #126 was caused by the compiler doing something funky - know we know its a bug we can eliminate the compiler as the root cause.

First, I can image this situation arising again, and it'd help to eliminate (some?) compiler optimizations as an issue.

Second, I can envision `#[trace]` being used with the Criterion crate - even through I've advocated minitrace using iai - wherein black-box is an option.

**Comment by andylokandy** [comment]
From my personal point of view, black_box will not affect the behavior of a program, but only the performance. And further, using `black_box` in benchmark code is sufficient to ensure the correctness of criterion.

**Comment by taqtiqa-mark** [comment]
The more I think about this the more I feel that by the time it lands it won't really be needed.  

I may have saved some time by wrapping the `fn` in question in black-box but I could have done that by hand.  Hopefully the phase of many bugs will shortly be passed and this is not longer a convenience.  Which makes this a catch-22 change - it'll only land at a point when it is no longer useful - after the proc-macro test suite is complete.

Closing.
